File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/value/impl/InnerAttValue.java
package ru.citeck.ecos.records3.record.op.atts.service.value.impl;

import ecos.com.fasterxml.jackson210.databind.JsonNode;
import ecos.com.fasterxml.jackson210.databind.node.ArrayNode;
import ecos.com.fasterxml.jackson210.databind.node.MissingNode;
import ecos.com.fasterxml.jackson210.databind.node.NullNode;
import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.records2.RecordConstants;
import ru.citeck.ecos.records2.graphql.meta.value.HasCollectionView;
import ru.citeck.ecos.records3.record.op.atts.service.value.AttValue;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;

public class InnerAttValue implements AttValue, HasCollectionView<InnerAttValue> {

    private final JsonNode value;

    public InnerAttValue(Object value) {
        this.value = Json.getMapper().convert(value, JsonNode.class);
    }

    public InnerAttValue(JsonNode value) {
        if (value == null || value instanceof MissingNode) {
            this.value = NullNode.getInstance();
        } else {
            this.value = value;
        }
    }

    @Override
    public Object getAtt(@NotNull String name) {

        JsonNode node = value.path(name);
        if (node.isMissingNode() || node.isNull()) {
            return null;
        }
        return new InnerAttValue(node);
    }

    @Override
    public Object as(@NotNull String type) {

        JsonNode node = value.path(RecordConstants.ATT_AS).path(type);
        if (node.isMissingNode() || node.isNull()) {
            return null;
        }

        return new InnerAttValue(node);
    }

    @Override
    public Collection<InnerAttValue> getCollectionView() {
        if (value instanceof ArrayNode) {
            List<InnerAttValue> result = new ArrayList<>();
            for (JsonNode val : value) {
                result.add(new InnerAttValue(val));
            }
            return result;
        }
        return Collections.singletonList(this);
    }

    @Override
    public boolean has(@NotNull String name) {

        JsonNode node = value.path(RecordConstants.ATT_HAS)
            .path(name)
            .path("?bool");

        if (node.isMissingNode() || node.isNull()) {
            return false;
        }
        return node.asBoolean();
    }

    @Override
    public String getDispName() {
        return getScalar(value, "?disp", JsonNode::asText);
    }

    @Override
    public String asText() {
        return getScalar(value, "?str", JsonNode::asText);
    }

    @Override
    public String getId() {
        return getScalar(value, "?id", JsonNode::asText);
    }

    @Override
    public Double asDouble() {
        return getScalar(value, "?num", JsonNode::asDouble);
    }

    @Override
    public Boolean asBool() {
        return getScalar(value, "?bool", JsonNode::asBoolean);
    }

    @Override
    public Object asJson() {
        return getScalar(value, "?json", n -> n);
    }

    private static <T> T getScalar(JsonNode node, String name, Function<JsonNode, T> mapper) {
        JsonNode scalar = node.path(name);
        if (scalar.isNull() || scalar.isMissingNode()) {
            return null;
        }
        return mapper.apply(scalar);
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/value/impl/InnerAttValue.kt
package ru.citeck.ecos.records3.record.op.atts.service.value.impl

import ecos.com.fasterxml.jackson210.databind.JsonNode
import ecos.com.fasterxml.jackson210.databind.node.ArrayNode
import ru.citeck.ecos.commons.json.Json
import ru.citeck.ecos.records2.RecordConstants
import ru.citeck.ecos.records3.record.op.atts.service.value.AttValue
import ru.citeck.ecos.records3.record.op.atts.service.value.HasCollectionView
import java.util.*

class InnerAttValue(value: Any?) : AttValue, HasCollectionView<InnerAttValue> {

    companion object {

        private fun <T> getScalar(node: JsonNode, name: String, mapper: (JsonNode) -> T): T? {
            val scalar = node.path(name)
            return if (scalar.isNull || scalar.isMissingNode) {
                null
            } else {
                mapper.invoke(scalar)
            }
        }
    }

    private val value: JsonNode = Json.mapper.toJson(value)

    override fun getAtt(name: String): Any? {
        val node = value.path(name)
        return if (node.isMissingNode || node.isNull) {
            null
        } else {
            InnerAttValue(node)
        }
    }

    override fun getAs(type: String): Any? {
        val node: JsonNode = value.path(RecordConstants.ATT_AS).path(type)
        return if (node.isMissingNode || node.isNull) {
            null
        } else {
            InnerAttValue(node)
        }
    }

    override fun getCollectionView(): Collection<InnerAttValue> {
        if (value is ArrayNode) {
            val result: MutableList<InnerAttValue> = ArrayList()
            for (elem in value) {
                result.add(InnerAttValue(elem))
            }
            return result
        }
        return listOf(this)
    }

    override fun has(name: String): Boolean {
        val node: JsonNode = value.path(RecordConstants.ATT_HAS)
            .path(name)
            .path("?bool")
        return if (node.isMissingNode || node.isNull) {
            false
        } else {
            node.asBoolean()
        }
    }

    override fun getDisplayName(): String? {
        return getScalar(value, "?disp") { it.asText() }
    }

    override fun asText(): String? {
        return getScalar(value, "?str") { it.asText() }
    }

    override fun getId(): String? {
        return getScalar(value, "?id") { it.asText() }
    }

    override fun asDouble(): Double? {
        return getScalar(value, "?num") { it.asDouble() }
    }

    override fun asBoolean(): Boolean? {
        return getScalar(value, "?bool") { it.asBoolean() }
    }

    override fun asJson(): Any? {
        return getScalar(value, "?json") { it }
    }
}
